org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/helper/WhileToForEach.java

```diff
@@ -0,0 +1,275 @@
+ *     Carsten Hammer
+ *******************************************************************************/
+package org.eclipse.jdt.internal.corext.fix.helper;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.text.edits.TextEditGroup;
+
+import org.eclipse.jdt.core.dom.AST;
+import org.eclipse.jdt.core.dom.ASTNode;
+import org.eclipse.jdt.core.dom.CompilationUnit;
+import org.eclipse.jdt.core.dom.EnhancedForStatement;
+import org.eclipse.jdt.core.dom.Expression;
+import org.eclipse.jdt.core.dom.IBinding;
+import org.eclipse.jdt.core.dom.IMethodBinding;
+import org.eclipse.jdt.core.dom.ITypeBinding;
+import org.eclipse.jdt.core.dom.MethodInvocation;
+import org.eclipse.jdt.core.dom.ParameterizedType;
+import org.eclipse.jdt.core.dom.SimpleName;
+import org.eclipse.jdt.core.dom.SimpleType;
+import org.eclipse.jdt.core.dom.SingleVariableDeclaration;
+import org.eclipse.jdt.core.dom.ThisExpression;
+import org.eclipse.jdt.core.dom.Type;
+import org.eclipse.jdt.core.dom.VariableDeclarationFragment;
+import org.eclipse.jdt.core.dom.VariableDeclarationStatement;
+import org.eclipse.jdt.core.dom.WhileStatement;
+import org.eclipse.jdt.core.dom.rewrite.ASTRewrite;
+
+import org.eclipse.jdt.internal.common.HelperVisitor;
+import org.eclipse.jdt.internal.common.ReferenceHolder;
+import org.eclipse.jdt.internal.corext.dom.ASTNodes;
+import org.eclipse.jdt.internal.corext.dom.Bindings;
+import org.eclipse.jdt.internal.corext.fix.CompilationUnitRewriteOperationsFixCore.CompilationUnitRewriteOperation;
+import org.eclipse.jdt.internal.corext.fix.ConvertLoopOperation;
+import org.eclipse.jdt.internal.corext.fix.UseIteratorToForLoopFixCore;
+import org.eclipse.jdt.internal.corext.refactoring.structure.CompilationUnitRewrite;
+
+/**
+ * Find: while (it.hasNext()){ System.out.println(it.next()); }
+ *
+ * Rewrite: for(Object o:collection) { System.out.println(o); });
+ *
+ */
+public class WhileToForEach extends AbstractTool<WhileLoopToChangeHit> {
+
+	@Override
+	public void find(UseIteratorToForLoopFixCore fixcore, CompilationUnit compilationUnit,
+			Set<CompilationUnitRewriteOperation> operations, Set<ASTNode> nodesprocessed) {
+		ReferenceHolder<ASTNode, WhileLoopToChangeHit> dataholder= new ReferenceHolder<>();
+		Map<ASTNode, WhileLoopToChangeHit> operationsMap= new LinkedHashMap<>();
+		WhileLoopToChangeHit emptyHit= new WhileLoopToChangeHit();
+		HelperVisitor.callVariableDeclarationStatementVisitor(Iterator.class, compilationUnit, dataholder, nodesprocessed, (init_iterator, holder_a) -> {
+			List<Object> computeVarName= computeVarName(init_iterator);
+			if (computeVarName != null) {
+				HelperVisitor.callWhileStatementVisitor(init_iterator.getParent(), dataholder, nodesprocessed, (whilestatement, holder) -> {
+					String name= computeNextVarname(whilestatement);
+					if (computeVarName.get(0).equals(name)) {
```

Comment:
Is this the usual way to check we are refering to the right variable? Should we somehow consider multiple variable cases?